File: python/EditDistanceTwoStrings.py

from random import *

def dpEditDistance(s1, s2, M, N):
    result = []
    for m in range(0, M + 1):
        result.append([(m + n) * (m == 0 or n == 0) for n in range(0, N + 1) ])

    for i in range(1, M + 1):
        for j in range(1, N + 1):
            insertion = result[i][j - 1] + 1
            deletion = result[i - 1][j] + 1
            matchOrMismatch = result[i - 1][j - 1] + 1 * (s1[i - 1 : i] != s2[j - 1 : j])

            result[i][j] = min(insertion, deletion,matchOrMismatch)

    for r in result:
        print(r)

    return result[M][N]

'''
s1, s2 = input(), input()
print(dpEditDistance(s1, s2, len(s1), len(s2)))
'''

'''
while True:
    len1 = randint(1,100)
    len2 = randint(1,100)
    s1 = ''.join([chr(randint(97,122)) for i in range(0, len1)])
    s2 = ''.join([chr(randint(97,122)) for j in range(0, len2)])

    print(s1)
    print(s2)
    print(dpEditDistance(s1, s2, len(s1), len(s2)))
    print('-----------------------------------------------')
'''


f = open('../sample/5_3_edit_distance.in')
s1 = f.readline()
s2 = f.readline()
print(dpEditDistance(s1, s2, len(s1), len(s2)))

File: c++/EditDistanceTwoStrings.cpp
#include "stdafx.h"
#include "EditDistanceTwoStrings.h"

using namespace std;

int dpEditDistanceTwoStrings(string s1, string s2, int M, int N)
{
	int** result = new int*[M + 1];
	for (int i = 0; i < M + 1; i++) {
		result[i] = new int[N + 1];
	}

	for (int i = 0; i < M + 1; i++) {
		for (int j = 0; j < N + 1; j++) {
			if (i == 0 || j == 0) {
				result[i][j] = i + j;
			}
			else {
				int insertion = result[i][j - 1] + 1;
				int deletion = result[i - 1][j] + 1;
				int mismatchOrMatch = result[i - 1][j - 1] + (s1.at(i - 1) == s2.at(j - 1) ? 0 : 1);

				result[i][j] = min({ insertion, deletion, mismatchOrMatch });
			}
		}
	}

	/*for (int i = 0; i < M + 1; i++) {
		for (int j = 0; j < N + 1; j++) {
			cout << result[i][j];
		}
		cout << endl;
	}*/

	return result[M][N];

}

EditDistanceTwoStrings::EditDistanceTwoStrings()
{
	/*string s1, s2;
	cin >> s1 >> s2;

	cout << dpEditDistanceTwoStrings(s1, s2, s1.size(), s2.size()) << endl;*/

	string s1, s2;
	ifstream file;
	file.open("./sample/5_3_edit_distance.in");
	if (!file) {
		cout << "cannot open a file!" << endl;
		exit(-1);
	}
	file >> s1 >> s2;
	cout << dpEditDistanceTwoStrings(s1, s2, s1.size(), s2.size()) << endl;

}


EditDistanceTwoStrings::~EditDistanceTwoStrings()
{
}
